readthedocs/rtd_tests/tests/test_api.py

```diff
@@ -978,6 +984,94 @@ def test_github_invalid_webhook(self, trigger_build):
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.data['detail'], 'Unhandled webhook event')
 
+    def test_github_invalid_payload(self, trigger_build):
+        client = APIClient()
+        integration = Integration.objects.create(
+            project=self.project,
+            integration_type=Integration.GITHUB_WEBHOOK,
+            secret=get_secret(),
+        )
+        wrong_signature = '1234'
+        self.assertNotEqual(integration.secret, wrong_signature)
+        headers = {
+            GITHUB_EVENT_HEADER: GITHUB_PUSH,
+            GITHUB_SIGNATURE_HEADER: wrong_signature,
+        }
+        resp = client.post(
+            '/api/v2/webhook/github/{}/'.format(self.project.slug),
+            self.github_payload,
+            format='json',
+            **headers
+        )
+        self.assertEqual(resp.status_code, 403)
+        self.assertEqual(resp.data['detail'], 'Payload not valid')
+
+    def test_github_valid_payload(self, trigger_build):
+        client = APIClient()
+        payload = '{"ref":"master"}'
+        integration = Integration.objects.create(
+            project=self.project,
+            integration_type=Integration.GITHUB_WEBHOOK,
+            secret=get_secret(),
+        )
+        digest = hmac.new(
+            integration.secret.encode(),
+            msg=payload.encode(),
+            digestmod=hashlib.sha1
+        ).hexdigest()
+        headers = {
+            GITHUB_EVENT_HEADER: GITHUB_PUSH,
+            GITHUB_SIGNATURE_HEADER: 'sha1=' + digest,
+        }
+        resp = client.post(
+            '/api/v2/webhook/github/{}/'.format(self.project.slug),
+            json.loads(payload),
+            format='json',
+            **headers
+        )
+        self.assertEqual(resp.status_code, 200)
+
+    def test_github_empty_signature(self, trigger_build):
+        client = APIClient()
+        headers = {
+            GITHUB_EVENT_HEADER: GITHUB_PUSH,
+            GITHUB_SIGNATURE_HEADER: '',
+        }
+        integration = Integration.objects.create(
+            project=self.project,
+            integration_type=Integration.GITHUB_WEBHOOK,
+            secret=get_secret(),
+        )
+        resp = client.post(
+            '/api/v2/webhook/github/{}/'.format(self.project.slug),
+            self.github_payload,
+            format='json',
+            **headers
+        )
+        self.assertEqual(resp.status_code, 403)
+        self.assertEqual(resp.data['detail'], 'Payload not valid')
+
+    def test_github_skip_signature_validation(self, trigger_build):
+        client = APIClient()
+        payload = '{"ref":"master"}'
+        integration = Integration.objects.create(
+            project=self.project,
+            integration_type=Integration.GITHUB_WEBHOOK,
+            secret='',
+        )
+        self.assertFalse(integration.secret)
+        headers = {
+            GITHUB_EVENT_HEADER: GITHUB_PUSH,
+            GITHUB_SIGNATURE_HEADER: 'skipped',
```

Comment:
Is this a special value, or is this test testing for the case where we previously didn't have a value for secret and won't validate the secret digest GitHub sends because of that?

Reply:
It's validating the old webhooks were we don't have a secret, yeah